Replaceable single entry point for parser

c42f opened the pull request:
This is a refactoring to allow the flisp parser to be replaced at runtime with some other parser as I've started to do in https://github.com/JuliaLang/FancyDiagnostics.jl/pull/4.

I've largely separated this into logically distinct commits so it may be worth looking at those in isolation while reviewing. (The main place where the separation isn't quite clean is in some fluctuation of the parsing function API.) To summarize the implementation:

* All parsing goes through a `jl_parse` interface which lives in a new file frontend.c. The idea of this new file is to eventually contain all frontend-related C code, some of which should move from ast.c over time.
* `jl_parse` currently has a C ABI and the implementation can be swapped out with `jl_set_parser`. I'm somewhat unsure about whether this is best compared to maintaining a global parser object in julia, but it seems potentially easier for solving future bootstrap issues.
* The flisp parser now has a single entry point `jl_fl_parse` which is compatible with `jl_parse`. This and other flisp entry points still live in `ast.c`.
* I have moved the implementation of `include` into Julia, which allowed the `mapexpr` handling in `jl_load_rewrite_file_string` and `jl_load_rewrite` to be removed and implemented in Julia instead, and also allows `jl_parse_eval_all` to be only used within bootstrap and for implementing the older `jl_load` and `jl_load_file_string` C API (actually these could be changed to call `Base.include` in future).
* As part of moving `include`, I've introduced two small internal functions `Meta.parseall` and `Meta.parseatom` to accompany `Meta.parse`. I've used this naming as I think this is what @JeffBezanson preferred; these can be further elaborated in #34715 and made public there.
* To preserve nicer stack traces from #33087, I've added `_simplify_include_frames` in order to not show users all the `include` implementation details which were previously hidden in the C code `jl_parse_eval_all`.  This only applies when `process_backtrace` is called to simplify the backtrace for printing.

### Future TODOs

Clearly this isn't the end-game for the parser API, but I think it moves in a useful direction and I'd like to get something merged here to avoid the work going stale.

A big unresolved issue is how the parser can communicate fine-grained location information to macro expansion, lowering, etc, and how that will be preserved through those passes. Macro expansion is particularly interesting as any naive approach will break user code! Eventually we want to preserve location info in a CSTParser-like data structure, but for now `jl_parse` returns a normal `Expr`.

I felt like `jl_parse` should accept rich options in principle (for example, to support communicating parser state when reparsing an arbitrary piece of text, and other such things). This is partly why it accepts a `jl_value_t` for the options rather than an integer I had in the earlier patches here. It remains to be seen exactly what options we might need, however.

Also I should say that bootstrap and the full removal of the flisp frontend is completely unexplored so far.

### Original description:

This is the start of a replacement for #35149, for discussion. The goals are
* [x] Have a single point of entry to the parser
* [x] Use this internally from everywhere we parse code
* [x] Make it so we can swap this out at runtime
* [x] Make sure it hooks up nicely to a user-facing API (Probably `Meta.parse`; cf discussion at https://github.com/JuliaLang/julia/pull/34715)
* [x] Try it out with CSTParser
* [x] Try moving `parse_eval_all` into loading.jl

~~I may leave this PR as a WIP (while I make a mess learning how everything slots together) and pull smaller logical pieces out as separate PRs for detailed review.~~

c42f committed:
Swappable entry point Core._parse for Julia parser

Add Core._parse binding as the internal entry point for parsing of Julia
code. This is called by both C and Julia code which wishes to parse
Julia. For now, it points at the flisp parser by default.

Refactor include() implementation to be done in Julia code, and parsing
rearrangement:

* Implement include() and include_string() in Julia. This also reunifies
  the versions of include() in MainInclude and Base which had started
  diverging.
* Add internal Core.Compiler.fl_parse and set this as the default value
  of the Core._parse binding.
* Use Core._parse from Meta
* Add `Meta.parseall()` for top-level parsing. For now as an internal
  function, but this will likely be made public shortly.

Refactoring in C runtime code:

* Remove flisp code from jl_parse_eval_all. This makes the top level
  parse-lower-eval loop independent of flisp internals.
* Remove jl_parse_eval_all from use as the include implementation (still
  used during bootstrap). This also allowed the removal of mapexpr
  handling from the C code.
* Keep jl_load and jl_load_file_string from julia.h unchanged for
  compatibility with the existing C API. (No julia code `ccall`s
  these anymore)
* Unify flisp parsing to always parse from string rather than directly
  from file.

c42f committed:
Add clang compilation database to gitignore

c42f committed:
Some comments about how to use htable_t

c42f committed:
In errorshow UI, filter out frames for _include implementation

This simplifies stack traces for include() which were made much more
complex by moving jl_parse_eval_all to the Julia layer.

c42f commented:
Ok, after some more cleanup and chasing down of a few broken tests I think this is ready for a closer look @Keno.

@vtjnash as you can see this has a single entry point for the parser in `jl_parse`. I've kept C ABI for now as it seemed this might be easier for bootstrap when we try to remove the flisp parser one day. If it's realistic to bootstrap with a Julia ABI and an approach more similar to `jl_set_typeinf_func` that alternative would also fine by me. At this stage I haven't experimented with how bootstrap could work with a Julia parser so I don't know what's best there.

BTW it's certain that this isn't the end game for the frontend, but I've tried to get to a reasonable intermediate state and document where it's at.

vtjnash commented on src/toplevel.c:
```diff
@@ -885,27 +885,139 @@ JL_DLLEXPORT jl_value_t *jl_infer_thunk(jl_code_info_t *thk, jl_module_t *m)
+// Synchronously read content of entire file into a julia String
+static jl_value_t *jl_file_content_as_string(jl_value_t *filename)
+{
+    const char* fname = jl_string_data(filename);
```
```suggestion
    const char *fname = jl_string_data(filename);
```

vtjnash reviewed (commented):
Seems reasonably straightforward now—just some small changes.

vtjnash commented on base/loading.jl:
```diff
@@ -1123,7 +1142,28 @@ The optional first argument `mapexpr` can be used to transform the included code
+    tls = task_local_storage()
+    tls[:SOURCE_PATH] = path
+    try
+        include_string(mapexpr, mod, code, path)
```
```suggestion
        return include_string(mapexpr, mod, code, path)
```
(because I like helping the eye find the result statement)

vtjnash commented on base/loading.jl:
```diff
@@ -1081,14 +1081,33 @@ The optional first argument `mapexpr` can be used to transform the included code
+        result = nothing
+        line_and_ex = Expr(:toplevel, loc, nothing)
+        for ex in ast.args
+            if ex isa LineNumberNode
```
so this removes our hacky globals
```
      jl_lineno = 0;
      jl_filename = jl_string_data(filename);
```
I guess that's a good thing, it's just the end of an era, like discarding of a well-worn shoe.

c42f replied:
Haha! Yes it's going in that direction, though it doesn't quite remove them: the internal machinery will still set them, and their use in extreme circumstances (`jl_critical_error`) will still work the same as it currently does (and be exactly as unthreadsafe as it currently is).

vtjnash reviewed (commented):
Seems reasonably straightforward now—just some small changes.

c42f commented on src/ast.c:
```diff
@@ -25,6 +27,8 @@ extern "C" {
 #pragma warning(disable:4335)
 #endif
 
+// TODO: Move non-flisp stuff into frontend.c (symbol init + macro expansion etc)
```
Sure, I'm happy to move the symbol init in here right away. (The only reason not to do this was to avoid excessive code movement in this PR.)

c42f replied:
Macro expansion would be fine too, except from the flisp bits which do `esc` handling.

c42f commented on src/support/htable.h:
```diff
@@ -6,11 +6,21 @@
 #define HT_N_INLINE 32
 
 #include "analyzer_annotations.h"
+#include <stddef.h>
```
Required for `size_t`, to have this header be valid in its own right, independently of including other things. Having the headers be self contained comes in handy if you're using clangd to resolve symbols.

c42f commented on stdlib/Distributed/test/distributed_exec.jl:
```diff
@@ -1481,12 +1481,19 @@ let
+        function test_include_fails_to_open_file(fname)
+            try
+                include(fname)
+                false
```
Leftover from refactoring. Will remove.

c42f commented on test/loading.jl:
```diff
@@ -570,7 +570,11 @@ end
 # normalization of paths by include (#26424)
-@test_throws ErrorException("could not open file $(joinpath(@__DIR__, "notarealfile.jl"))") include("./notarealfile.jl")
+@test begin
+    exc = try include("./notarealfile.jl") ; catch exc ; exc ; end
```
When run with the standard test harness, `include("loading.jl")` is eventually run on one of the test workers.  So the effective working directory for this `include` should be the `test` directory.

c42f commented on base/errorshow.jl:
```diff
@@ -750,6 +750,42 @@ function is_kw_sorter_name(name::Symbol)
+    i = length(trace)
+    first_ignored = nothing
+    while i >= 1
+        frame,_ = trace[i]
```
Yes this works, but I think being explicit about discarding stuff is more helpful than harmful.

vtjnash replied:
```suggestion
        frame, _ = trace[i]
```

c42f commented on base/loading.jl:
```diff
@@ -1081,14 +1081,33 @@ The optional first argument `mapexpr` can be used to transform the included code
+            # Wrap things to be eval'd in a :toplevel expr to carry line
+            # information as part of the expr.
+            line_and_ex.args[2] = ex
+            # Macro expand, lower, and evaluate successive top-level statements
+            # in the latest world.
+            result = invokelatest(Core.eval, mod, line_and_ex)
```
Sure, I'm happy to remove this.

But why did `jl_parse_eval_all` previously set the world age as part of its evaluation loop?

c42f commented on base/loading.jl:
```diff
@@ -1081,14 +1081,33 @@ The optional first argument `mapexpr` can be used to transform the included code
+    catch exc
+        # TODO: Now that stacktraces are more reliable we should remove
+        # LoadError and expose the real error type directly.
+        rethrow(LoadError(String(loc.file), loc.line, exc))
```
Thanks very much. Will fix.

c42f replied:
```suggestion
        rethrow(LoadError(filename, loc.line, exc))
```

c42f commented on base/meta.jl:
```diff
@@ -149,6 +149,29 @@ struct ParseError <: Exception
+    options = values(options)
+    if length(options) == 1 && haskey(options, :rule)
+        # Pass rule as a symbol to simplify jl_fl_parse
+        options = options.rule
```
Thanks, I think I'll just simplify this and pass the rule around as a symbol everywhere. I was trying to add more general options but it's unclear they will be needed and it's clear that `Base` won't know about them in this iteration.

c42f commented on src/ast.c:
```diff
@@ -2,7 +2,9 @@
 /*
   AST
-  interface to front-end, obtains and translates syntax trees
+  interface to flisp front-end, obtains and translates syntax trees
+
+  This file contains all code which interacts with the flisp frontend.
```
Why delete this comment? Searching for `fl_apply`, you'll find that it's technically correct: this file does contain all code which calls into the flisp frontend.

Perhaps I misread your comment in `frontend.c`.  Do you mean that we should just put frontend.c back into ast.c?

c42f replied:
Oops. I see I was attributing the comment to the wrong file (reading backward :man_facepalming:). Carry on, nothing to see here!

c42f commented on src/ast.c:
```diff
@@ -774,185 +785,57 @@ static value_t julia_to_scm_(fl_context_t *fl_ctx, jl_value_t *v)
-JL_DLLEXPORT jl_value_t *jl_parse_all(const char *str, size_t len, const char *filename, size_t filename_len)
+// Parse string `content` starting at 0-based index `offset` attributing the
+// content to `filename`. Return an svec of (parse_result, final_pos)
+JL_DLLEXPORT jl_value_t *jl_fl_parse(const char* text, size_t text_len,
+                                     const char* filename, size_t filename_len,
```
Thanks.  Old habits die hard.

c42f commented on src/ast.c:
```diff
@@ -774,185 +785,57 @@ static value_t julia_to_scm_(fl_context_t *fl_ctx, jl_value_t *v)
-        fl_gc_handle(fl_ctx, &t);
-        ast = fl_applyn(fl_ctx, 2, symbol_value(symbol(fl_ctx, "jl-parse-all")), t, f);
-        fl_free_gc_handles(fl_ctx, 1);
+    value_t fl_text = cvalue_static_cstrn(fl_ctx, text, text_len);
+    value_t fl_filename = cvalue_static_cstrn(fl_ctx, filename, filename_len);
```
Phew, I did not know about this. Thanks for the thorough review.

c42f commented on src/frontend.c:
```diff
@@ -0,0 +1,56 @@
+                                  const char* filename, size_t filename_len,
+                                  size_t offset, jl_value_t* options)
+{
+    return (*jl_current_parser)(text, text_len, filename, filename_len, offset, options);
```
Ok then, let's do it that way, it's simpler anyway. We can worry about bootstrap later (if indeed it makes any difference to this).

c42f commented on base/meta.jl:
```diff
@@ -171,19 +194,11 @@ julia> Meta.parse("x = 3, y = 5", 5)
-              (Ptr{UInt8}, Csize_t, Int32, Int32),
-              bstr, sizeof(bstr), pos-1, greedy ? 1 : 0)
-    end
+    ex, pos = _jl_parse(str, "none", pos; rule=greedy ? :statement : :atom)
```
Will clean up as part of simplifying `rule` handling.

c42f commented:
This has been updated to move almost the whole parser replacement interface into Juila, with a few tiny shims in `Core.Compiler` to make this possible. I think the resulting interface is a lot more satisfying and hopefully somewhat more future proof.

One thing I find particularly satisfying is that the ownership of the text buffer can be safely communicated through the interface. So if the parser wants to return an error message which references the text buffer, and that buffer is a `String` (or other Julia object), there's no need to make an entire copy of the buffer. With the buffer communicated as a `Ptr{UInt8}` in the previous setup, the only real option was to copy the entire buffer if you wanted to preserve it.

Regarding the ability to freeze the parser in a certain world, I added a `_apply_in_world` builtin so that this could be done from Julia code. This generalizes `_apply_latest` and could in fact replace it, though I didn't go that far here. After adding `_apply_in_world`, I realized that it's awkward to make this part of the interface through which the parser is called. Rather, the builtin allows the parser itself to make the decision about whether it runs in a given world, and what world that needs to be.

vtjnash commented on base/stacktraces.jl:
```diff
@@ -268,22 +268,29 @@ function show(io::IO, frame::StackFrame; full_path::Bool=false)
+            return def
+        else
+            return (def::Method).module
+            return def.module
```
```suggestion
```

vtjnash commented on base/stacktraces.jl:
```diff
@@ -268,22 +268,29 @@ function show(io::IO, frame::StackFrame; full_path::Bool=false)
+            return def.module
+        end
+    else
+        # Bug: currently the module is not available for inlined frames and
```
```suggestion
        # the module is not always available (common reasons include inlined frames and frames arising from the interpreter)
```
Not a bug for debug information to be incomplete

c42f replied:
Fair enough, will change. It's not a bug itself (though can easily lead to bugs).

vtjnash commented on base/stacktraces.jl:
```diff
@@ -268,22 +268,29 @@ function show(io::IO, frame::StackFrame; full_path::Bool=false)
-    end
-
-    return result
+    parentmodule(frame) == m
```
```suggestion
    return parentmodule(frame) === m
```

c42f replied:
Happy to change this, but curious what the reasoning is as I thought they're equivalent here. Is it because you think `===` is semantically clearer?

vtjnash commented on src/ast.c:
```diff
@@ -1277,6 +1161,85 @@ JL_DLLEXPORT jl_value_t *jl_expand_stmt(jl_value_t *expr, jl_module_t *inmodule)
+    jl_value_t *result = jl_apply(args, 5);
+    ptls->world_age = last_age;
+    args[0] = result; // root during error checks below
+    if (!jl_is_svec(result)) {
+        jl_type_error("jl_parse", (jl_value_t*)jl_simplevector_type, result);
+    }
+    else if (jl_svec_len(result) != 2 || !jl_is_expr(jl_svecref(result, 0)) ||
+             !jl_is_long(jl_svecref(result, 1))) {
+        jl_error("Result from parser should be `svec(a::Expr, b::Int)`");
+    }
```
Somewhat more canonical form:
```suggestion
JL_TYPECHK(parse, simplevector, result);
if (jl_svec_len(result) != 2)
    jl_error("Result from parser should be `svec(a::Expr, b::Int)`");
JL_TYPECHK(parse, expr, jl_svecref(result, 0));
JL_TYPECHK(parse, long, jl_svecref(result, 1));
```

c42f replied:
Oh right, thanks.

I tried something like this but `JL_TYPECHK(parse, svec, result)` didn't work, and I didn't realize that `simplevector` could be substituted.

c42f commented on stdlib/Distributed/test/distributed_exec.jl:
```diff
@@ -1483,12 +1483,19 @@ let
+        function test_include_fails_to_open_file(fname)
+            try
+                include(fname)
+                false
```
```suggestion
```

vtjnash reviewed (commented):
Seems good to me thus far. Slightly unfortunate to be causing Base now to require the existence of `Core.Compiler`, so perhaps we should put the bare minimum of those symbols into `boot.jl`, but not a show-stopper towards merging this.

c42f commented:
> Slightly unfortunate to be causing Base now to require the existence of `Core.Compiler`, so perhaps we should put the bare minimum of those symbols into `boot.jl`

How about I move the entire content of parsing.jl into boot.jl? It's only four new symbols.

c42f commented:
Ok, I've addressed all suggested cleanups, and moved the `Core.Compiler.parse` shim to `Core.parse`, to break an explicit dependency from `Base` to `Core.Compiler`. Also I just deleted `Core.Compiler.set_parser`, as a simple `Core.eval(:(_parser = $whatever))` will suffice for now and we can decide whether we actually want any API for this later.

c42f commented:
Rebased to fix the mess of commits and clean up a few things. Notable cleanup:
* There's now only a single binding `Core._parse` which is added to Core.
* I removed the `_apply_in_world` builtin from here - will move to a separate PR.

I think this is in a good state now so I plan to merge it later today unless there's anything further.

JeffBezanson commented on base/boot.jl:
```diff
@@ -740,4 +740,19 @@ Unsigned(x::Union{Float32, Float64, Bool}) = UInt(x)
+# `Int`.
+#
+# The internal jl_parse which will call into Core._parse if not `nothing`.
+eval(Core, :(_parse = nothing))
```
Isn't this file already in `Core`?

c42f replied:
Oops! Will fix, thanks.

JeffBezanson commented:
Looking good!
